src/core/dataset.rs

```diff
@@ -125,13 +131,49 @@ impl<T: Number, U: Number> Dataset<T, U> for VecVec<T, U> {
     fn query_to_one(&self, query: &[T], index: usize) -> U {
         (self.metric)(query, &self.data[index])
     }
+
+    fn reorder(&mut self, indices: &Vec<usize>) {
+        let n = indices.len();
+
+        assert!(n == self.cardinality());
+
+        // The "source index" represents the index that we hope to swap to
+        let mut source_index: usize;
+
+        // INVARIANT: After each iteration of the loop, the elements of the
+        // subarray [0..i] are in the correct position.
+        for i in 0..n-1 {
+            source_index = indices[i];
+
+            // If the element at is already at the correct position, we can
+            // just skip.
+            if source_index != i {
+                // Here we're essentially following the cycle. We *know* by
+                // the invariant that all elements to the left of i are in
+                // the correct position, so what we're doing is following
+                // the cycle until we find an index to the right of i. Which,
+                // because we followed the position changes, is the correct
+                // index to swap.
+                while source_index < i {
+                    source_index = indices[source_index];
+                }
+
+                // We swap to the correct index. Importantly, this index is always
+                // to the right of i, we do not modify any index to the left of i.
+                // Thus, because we followed the cycle to the correct index to swap,
+                // we know that the element at i, after this swap, is in the correct
+                // position.
+                self.data.swap(source_index, i);
```

Comment:
Add a required method to the trait with the following signature:
```rust
fn swap(&mut self, i: usize, j: usize) -> Result<(), String>;
```

Then implement the method for this struct and then move the implementation of this method to be a default implementation is the trait itself.

Reply:
For `VecVec`: What should the failure of a `swap` look like? The default behavior for `Vec::swap` when an index is out of bounds is to just panic. Should `VecVec::swap` do it's own bounds check `Err` accordingly or just let it panic?

Reply:
It's fine to panic. You can change the signature to not return anything.